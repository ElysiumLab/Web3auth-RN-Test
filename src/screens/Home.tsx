import { useState } from "react";
import { SafeAreaView, View, Text, Button, StyleSheet } from "react-native";
import { Wallet } from "ethers";

// Usefull for login with Web3Auth
import * as WebBrowser from "expo-web-browser";
import Web3Auth, {OPENLOGIN_NETWORK, TypeOfLogin} from "@web3auth/react-native-sdk";
// Get and create link
import Constants, { AppOwnership } from "expo-constants";
import * as Linking from "expo-linking";
import { Buffer } from "buffer";

import { FocusedStatusBar, HomeHeader } from "../components";
import { styles, COLORS } from "../styles";
import { CLIENT_ID, LOGIN_NAME, LOGIN_VERIFIER, LOGIN_TYPE, LOGIN_CLIENT_ID, LOGIN_PROVIDER, LOGIN_DOMAIN } from "@env";

global.Buffer = global.Buffer || Buffer;

// Create specific link for each platforms
const scheme = "web3authexposample";

const resolvedRedirectUrl =
  Constants.appOwnership == AppOwnership.Expo ||
  Constants.appOwnership == AppOwnership.Guest
    ? Linking.createURL("web3auth", {})
    : Linking.createURL("web3auth", { scheme: scheme });

export const Home = () => {
  const [wallet, setWallet] = useState(null);

  const web3auth = new Web3Auth(WebBrowser, {
    clientId: CLIENT_ID,
    network: OPENLOGIN_NETWORK.TESTNET,
    loginConfig: {
      jwt: {
        verifier: LOGIN_VERIFIER, // get it from web3auth dashboard for auth0 configuration
        typeOfLogin: LOGIN_TYPE,
        name: LOGIN_NAME,
        clientId: LOGIN_CLIENT_ID, // get it from auth0 dashboard
      }
    },
  });

  const login = async () => {
    try {
      const state = await web3auth.login({
        loginProvider: LOGIN_PROVIDER,
        redirectUrl: resolvedRedirectUrl,
        extraLoginOptions: {
          domain: LOGIN_DOMAIN, // domain of your auth0 app
          verifierIdField: "sub", // The field in jwt token which maps to verifier id.
        },
      });

      const wallet = new Wallet(state.privKey);

      setWallet(wallet);
    } catch (e) {
      console.error(e);
    }
  };

  const logout = async () => {
    try {
      await web3auth.logout({ redirectUrl: resolvedRedirectUrl });

      setWallet(null);
    } catch (e) {
      console.error(e);
    }
  };

  return (
    <SafeAreaView style={styles.safeAreaContainer}>
      <FocusedStatusBar />

      <HomeHeader
        ethAddress={wallet?.address}
        redirectUrl={resolvedRedirectUrl}
        onLogin={login}
        onLogout={logout}
      />
    </SafeAreaView>
  );
};

export default Home;

const style = StyleSheet.create({
  content: {
    alignItems: "center",
    justifyContent: "center",
  },
});
