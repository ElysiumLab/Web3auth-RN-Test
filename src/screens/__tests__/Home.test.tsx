import React from "react";
import { render } from "@testing-library/react-native";

import Home from "../Home";

// jest.mock("expo-constants", () => {
//   const actualConstants = jest.requireActual("expo-constants");

//   return {
//     ...actualConstants,
//     Constants: {
//       appOwnership: "expo",
//     },
//   };
// });

jest.mock("expo-linking"),
  () => ({
    createURL: "test",
  });

jest.mock("../../components", () => {
  const actualHomeHeader = jest.requireActual(
    "../../components/headers/HomeHeader"
  );

  const { HomeHeader } = actualHomeHeader;

  return {
    FocusedStatusBar: jest.fn(),
    HomeHeader: HomeHeader,
  };
});

describe("<Home />", () => {
  it("renders correctly", () => {
    const tree = render(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
