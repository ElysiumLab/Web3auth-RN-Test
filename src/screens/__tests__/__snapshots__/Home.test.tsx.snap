// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#001F2D",
        "height": 100,
      }
    }
  >
    <View
      style={
        Object {
          "position": "absolute",
          "right": 10,
          "top": 10,
          "width": 60,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#0000FF",
            "justifyContent": "center",
            "minHeight": 30,
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "textAlign": "center",
              },
              Object {
                "color": "#FFFFFF",
                "fontSize": 16,
              },
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "bottom": 10,
          "position": "absolute",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 12,
          }
        }
      >
        Redirect url: web3authexposample
      </Text>
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 12,
          }
        }
      >
        Eth Address:
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
